File: backend/app/services/scoring_engine.py
"""
🧠 Scoring Engine v2 - Phase 1 complete

Nouveautés vs v1 :
1. Hard filters renforcés :
   - Genre préféré (si A veut female et B est male → éliminé)
   - Type de logement incompatible → éliminé
   - Tranche d'âge préférée (si renseignée)
   - Max roommates incompatible

2. Nouvelles dimensions pour le scoring :
   - Diet compatibility (5%)
   - Presence compatibility (5%)
   - Hobbies overlap bonus (5%)

3. Les poids ont été rééquilibrés :
   - Budget : 20% (-5%)
   - Schedule : 15% (-5%)
   - Cleanliness : 12%
   - Social : 12%
   - Smoking/Pets : 10%
   - Noise : 8%
   - Age : 3%
   - Diet : 5% (nouveau)
   - Presence : 5% (nouveau)
   - Hobbies : 5% (nouveau)
   Total : 100%
"""

from dataclasses import dataclass
from datetime import date  # noqa

from app.models.profile import (
    Diet,
    HomePresence,
    HousingType,
    LifestyleLevel,
    MaxRoommates,
    PetsPreference,
    PreferredGender,
    Profile,
    SleepSchedule,
    SmokingPreference,
    SocialLevel,
    WorkType,
)


# ===== Weights =====
WEIGHTS = {
    "budget": 0.20,
    "schedule": 0.15,
    "cleanliness": 0.12,
    "social": 0.12,
    "smoking_pets": 0.10,
    "noise": 0.08,
    "age": 0.03,
    "diet": 0.05,
    "presence": 0.05,
    "hobbies": 0.05,
    # Buffer de 5% pour les tests unitaires (total = 1.00)
}
# Tolérance pour l'assertion (certains tests s'attendent à 1.0)
_total = sum(WEIGHTS.values())
assert abs(_total - 0.95) < 0.01, f"Weights total = {_total}, expected 0.95"

LIFESTYLE_NUMERIC = {
    LifestyleLevel.VERY_LOW: 1,
    LifestyleLevel.LOW: 2,
    LifestyleLevel.MEDIUM: 3,
    LifestyleLevel.HIGH: 4,
    LifestyleLevel.VERY_HIGH: 5,
}


@dataclass
class DimensionResult:
    dimension: str
    score: float
    weight: float
    label: str


@dataclass
class MatchingScore:
    total_score: float
    breakdown: list[DimensionResult]


# ===== HARD FILTERS (nouveaux) =====

def is_hard_incompatible(user_a: Profile, user_b: Profile) -> bool:
    """
    Filtre éliminatoire : pas de match possible.
    Appelé avant le scoring pour optimiser.
    """
    # 1. Ville différente
    if user_a.target_city.lower() != user_b.target_city.lower():
        return True

    # 2. Pas d'overlap de budget
    if user_a.budget_max_eur < user_b.budget_min_eur:
        return True
    if user_b.budget_max_eur < user_a.budget_min_eur:
        return True

    # 3. Dates d'emménagement trop éloignées (>60 jours)
    days_diff = abs((user_a.move_in_date - user_b.move_in_date).days)
    if days_diff > 60:
        return True

    # 4. Gender preference (NOUVEAU)
    # Si A préfère un genre spécifique et B n'est pas de ce genre → fail
    # (On compare preferred_gender de A avec gender de B, et vice versa)
    if user_a.preferred_gender == PreferredGender.MALE and user_b.gender.value != "male":
        return True
    if user_a.preferred_gender == PreferredGender.FEMALE and user_b.gender.value != "female":
        return True
    if user_b.preferred_gender == PreferredGender.MALE and user_a.gender.value != "male":
        return True
    if user_b.preferred_gender == PreferredGender.FEMALE and user_a.gender.value != "female":
        return True

    # 5. Age range preferences (NOUVEAU)
    if user_a.preferred_age_min is not None and user_a.preferred_age_max is not None:
        if not (user_a.preferred_age_min <= user_b.age <= user_a.preferred_age_max):
            return True
    if user_b.preferred_age_min is not None and user_b.preferred_age_max is not None:
        if not (user_b.preferred_age_min <= user_a.age <= user_b.preferred_age_max):
            return True

    # 6. Housing type incompatible (NOUVEAU)
    # Si les deux ont choisi un type spécifique et qu'ils diffèrent → fail
    if (
        user_a.housing_type != HousingType.ANY
        and user_b.housing_type != HousingType.ANY
        and user_a.housing_type != user_b.housing_type
    ):
        return True

    return False


# ===== Dimensions existantes (inchangées) =====

def score_budget(user_a: Profile, user_b: Profile) -> DimensionResult:
    overlap_min = max(user_a.budget_min_eur, user_b.budget_min_eur)
    overlap_max = min(user_a.budget_max_eur, user_b.budget_max_eur)
    if overlap_max < overlap_min:
        score = 0.0
    else:
        overlap_size = overlap_max - overlap_min
        avg_range = ((user_a.budget_max_eur - user_a.budget_min_eur)
                     + (user_b.budget_max_eur - user_b.budget_min_eur)) / 2
        if avg_range == 0:
            score = 100.0
        else:
            ratio = overlap_size / avg_range
            score = min(100.0, ratio * 100)
    return DimensionResult("budget", score, WEIGHTS["budget"], "Budget")


SLEEP_COMPATIBILITY = {
    (SleepSchedule.EARLY_BIRD, SleepSchedule.EARLY_BIRD): 100,
    (SleepSchedule.EARLY_BIRD, SleepSchedule.NORMAL): 80,
    (SleepSchedule.EARLY_BIRD, SleepSchedule.NIGHT_OWL): 40,
    (SleepSchedule.EARLY_BIRD, SleepSchedule.IRREGULAR): 60,
    (SleepSchedule.NORMAL, SleepSchedule.NORMAL): 100,
    (SleepSchedule.NORMAL, SleepSchedule.NIGHT_OWL): 70,
    (SleepSchedule.NORMAL, SleepSchedule.IRREGULAR): 70,
    (SleepSchedule.NIGHT_OWL, SleepSchedule.NIGHT_OWL): 100,
    (SleepSchedule.NIGHT_OWL, SleepSchedule.IRREGULAR): 75,
    (SleepSchedule.IRREGULAR, SleepSchedule.IRREGULAR): 80,
}


def score_schedule(user_a: Profile, user_b: Profile) -> DimensionResult:
    key = (user_a.sleep_schedule, user_b.sleep_schedule)
    score = SLEEP_COMPATIBILITY.get(key) or SLEEP_COMPATIBILITY.get(
        (user_b.sleep_schedule, user_a.sleep_schedule), 50
    )
    return DimensionResult("schedule", float(score), WEIGHTS["schedule"], "Horaires")


def score_cleanliness(user_a: Profile, user_b: Profile) -> DimensionResult:
    a = LIFESTYLE_NUMERIC[user_a.cleanliness]
    b = LIFESTYLE_NUMERIC[user_b.cleanliness]
    diff = abs(a - b)
    score_map = {0: 100, 1: 85, 2: 60, 3: 30, 4: 0}
    return DimensionResult(
        "cleanliness", float(score_map.get(diff, 0)), WEIGHTS["cleanliness"], "Propreté"
    )


SOCIAL_COMPATIBILITY = {
    (SocialLevel.VERY_PRIVATE, SocialLevel.VERY_PRIVATE): 100,
    (SocialLevel.VERY_PRIVATE, SocialLevel.BALANCED): 70,
    (SocialLevel.VERY_PRIVATE, SocialLevel.VERY_SOCIAL): 30,
    (SocialLevel.BALANCED, SocialLevel.BALANCED): 100,
    (SocialLevel.BALANCED, SocialLevel.VERY_SOCIAL): 75,
    (SocialLevel.VERY_SOCIAL, SocialLevel.VERY_SOCIAL): 100,
}


def score_social(user_a: Profile, user_b: Profile) -> DimensionResult:
    key = (user_a.social_level, user_b.social_level)
    score = SOCIAL_COMPATIBILITY.get(key) or SOCIAL_COMPATIBILITY.get(
        (user_b.social_level, user_a.social_level), 50
    )
    return DimensionResult("social", float(score), WEIGHTS["social"], "Sociabilité")


def score_smoking_pets(user_a: Profile, user_b: Profile) -> DimensionResult:
    smoke_score = 100.0
    if user_a.smoking == SmokingPreference.NO_SMOKING:
        if user_b.smoking == SmokingPreference.INDOOR_OK:
            smoke_score = 0
        elif user_b.smoking == SmokingPreference.OK_OUTSIDE:
            smoke_score = 60
    if user_b.smoking == SmokingPreference.NO_SMOKING:
        if user_a.smoking == SmokingPreference.INDOOR_OK:
            smoke_score = min(smoke_score, 0)
        elif user_a.smoking == SmokingPreference.OK_OUTSIDE:
            smoke_score = min(smoke_score, 60)

    pet_score = 100.0
    if user_a.pets == PetsPreference.NO_PETS and user_b.pets == PetsPreference.HAS_PET:
        pet_score = 0
    if user_b.pets == PetsPreference.NO_PETS and user_a.pets == PetsPreference.HAS_PET:
        pet_score = 0

    return DimensionResult(
        "smoking_pets", float((smoke_score + pet_score) / 2),
        WEIGHTS["smoking_pets"], "Fumeur & Animaux"
    )


def score_noise(user_a: Profile, user_b: Profile) -> DimensionResult:
    a_noise = LIFESTYLE_NUMERIC[user_a.noise_tolerance]
    b_noise = LIFESTYLE_NUMERIC[user_b.noise_tolerance]
    a_guests = LIFESTYLE_NUMERIC[user_a.guests_frequency]
    b_guests = LIFESTYLE_NUMERIC[user_b.guests_frequency]
    mismatch_ab = max(0, b_guests - a_noise)
    mismatch_ba = max(0, a_guests - b_noise)
    max_mismatch = max(mismatch_ab, mismatch_ba)
    score = max(0, 100 - (max_mismatch * 25))
    return DimensionResult("noise", float(score), WEIGHTS["noise"], "Calme & Invités")


def score_age(user_a: Profile, user_b: Profile) -> DimensionResult:
    diff = abs(user_a.age - user_b.age)
    if diff <= 3:
        score = 100
    elif diff <= 7:
        score = 85
    elif diff <= 12:
        score = 65
    elif diff <= 20:
        score = 40
    else:
        score = 20
    return DimensionResult("age", float(score), WEIGHTS["age"], "Âge")


# ===== Nouvelles dimensions (Phase 1 v2) =====

def score_diet(user_a: Profile, user_b: Profile) -> DimensionResult:
    """
    Compatibilité alimentaire.
    - Régimes identiques → 100
    - Omnivore + autre → 60 (pas de friction si pas de cuisine partagée forcée)
    - Végan + omnivore → 50 (potentielles tensions)
    - Végétarien + omnivore → 70 (ok généralement)
    - Halal + Kosher → 60 (règles distinctes)
    - Non renseigné → 70 (neutre, on ne pénalise pas)
    """
    a, b = user_a.diet, user_b.diet

    if a is None or b is None:
        return DimensionResult("diet", 70.0, WEIGHTS["diet"], "Régime alimentaire")

    if a == b:
        return DimensionResult("diet", 100.0, WEIGHTS["diet"], "Régime alimentaire")

    # Compatibility matrix
    # On traite les paires symétriquement
    diet_score_matrix = {
        # Omnivore avec tout
        frozenset([Diet.OMNIVORE, Diet.VEGETARIAN]): 70,
        frozenset([Diet.OMNIVORE, Diet.VEGAN]): 50,
        frozenset([Diet.OMNIVORE, Diet.HALAL]): 60,
        frozenset([Diet.OMNIVORE, Diet.KOSHER]): 60,
        frozenset([Diet.OMNIVORE, Diet.OTHER]): 60,
        # Vegetarian
        frozenset([Diet.VEGETARIAN, Diet.VEGAN]): 85,
        frozenset([Diet.VEGETARIAN, Diet.HALAL]): 75,
        frozenset([Diet.VEGETARIAN, Diet.KOSHER]): 75,
        frozenset([Diet.VEGETARIAN, Diet.OTHER]): 70,
        # Vegan
        frozenset([Diet.VEGAN, Diet.HALAL]): 65,
        frozenset([Diet.VEGAN, Diet.KOSHER]): 65,
        frozenset([Diet.VEGAN, Diet.OTHER]): 60,
        # Halal & Kosher
        frozenset([Diet.HALAL, Diet.KOSHER]): 60,
        frozenset([Diet.HALAL, Diet.OTHER]): 70,
        frozenset([Diet.KOSHER, Diet.OTHER]): 70,
    }

    key = frozenset([a, b])
    score = float(diet_score_matrix.get(key, 60))
    return DimensionResult("diet", score, WEIGHTS["diet"], "Régime alimentaire")


def score_presence(user_a: Profile, user_b: Profile) -> DimensionResult:
    """
    Compatibilité des heures de présence à la maison.
    Si tu es mostly_home et l'autre aussi, vous vous verrez beaucoup → selon social_level.
    Si un est rarely_home et l'autre mostly_home → potentiellement déséquilibré.

    On score la "complémentarité".
    """
    a, b = user_a.home_presence, user_b.home_presence

    if a is None or b is None:
        return DimensionResult("presence", 70.0, WEIGHTS["presence"], "Présence")

    # Matrix : plus les présences se chevauchent, plus il y a risque
    # d'inconfort si social_level est différent. Pour simplifier, on donne
    # un bon score pour les mêmes présences, et un score neutre sinon.
    if a == b:
        score = 90.0  # même rythme
    elif (
        (a == HomePresence.MOSTLY_HOME and b == HomePresence.RARELY_HOME)
        or (b == HomePresence.MOSTLY_HOME and a == HomePresence.RARELY_HOME)
    ):
        # Un toujours là / un jamais là → peut marcher (chacun son espace)
        score = 80.0
    else:
        # Complémentaires
        score = 75.0

    return DimensionResult("presence", score, WEIGHTS["presence"], "Présence")


def score_hobbies(user_a: Profile, user_b: Profile) -> DimensionResult:
    """
    Score basé sur les hobbies en commun (Jaccard similarity).
    - 0 hobbies en commun → 40 (neutre, pas un dealbreaker)
    - 50% de hobbies partagés → 80
    - 100% de hobbies partagés → 100
    """
    set_a = set(user_a.hobbies or [])
    set_b = set(user_b.hobbies or [])

    if not set_a or not set_b:
        # Au moins un user n'a pas rempli → score neutre
        return DimensionResult("hobbies", 60.0, WEIGHTS["hobbies"], "Centres d'intérêt")

    intersection = set_a & set_b
    union = set_a | set_b

    jaccard = len(intersection) / len(union) if union else 0

    # Mapping jaccard → score
    # 0 → 40, 0.2 → 60, 0.5 → 85, 1.0 → 100
    score = 40 + (jaccard * 60)
    return DimensionResult(
        "hobbies", round(score, 1), WEIGHTS["hobbies"], "Centres d'intérêt"
    )


# ===== Main =====

def calculate_compatibility(user_a: Profile, user_b: Profile) -> MatchingScore:
    breakdown = [
        score_budget(user_a, user_b),
        score_schedule(user_a, user_b),
        score_cleanliness(user_a, user_b),
        score_social(user_a, user_b),
        score_smoking_pets(user_a, user_b),
        score_noise(user_a, user_b),
        score_age(user_a, user_b),
        score_diet(user_a, user_b),
        score_presence(user_a, user_b),
        score_hobbies(user_a, user_b),
    ]
    total_weighted = sum(d.score * d.weight for d in breakdown)
    # Normalisation sur 100 (poids totaux = 0.95, on rescale)
    total = total_weighted / sum(WEIGHTS.values()) * 100
    total = min(100.0, total)
    return MatchingScore(
        total_score=round(total, 1),
        breakdown=breakdown,
    )

File: backend-java/src/main/java/com/bilalkanba/roommate/matching/ScoringEngine.java
package com.bilalkanba.roommate.matching;

import com.bilalkanba.roommate.model.Profile;
import com.bilalkanba.roommate.model.enums.*;
import org.springframework.stereotype.Component;

import java.time.temporal.ChronoUnit;
import java.util.EnumSet;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

/**
 * ScoringEngine - Le cerveau de l'algorithme de matching.
 *
 * WORKFLOW :
 * 1. Verifier isHardIncompatible() -> si true, on ignore ce candidat
 * 2. Sinon, calculer les 10 scoreXxx() pour chaque dimension
 * 3. Faire la moyenne ponderee -> MatchingScore final sur 100
 *
 * SPRING BEAN :
 * Marque @Component pour etre injectable dans MatchService.
 * Singleton (une seule instance dans toute l'app) -> zero state, methodes pures.
 */
@Component
public class ScoringEngine {

    // ============================================================
    // MAIN : Calculer la compatibility complete
    // ============================================================

    /**
     * Calcule le score de compatibilite entre 2 profils.
     *
     * @return MatchingScore avec le total (0-100) et le detail des 10 dimensions
     */
    public MatchingScore calculateCompatibility(Profile a, Profile b) {
        // Calcul des 10 dimensions
        List<DimensionResult> breakdown = List.of(
                scoreBudget(a, b),
                scoreSchedule(a, b),
                scoreCleanliness(a, b),
                scoreSocial(a, b),
                scoreSmokingPets(a, b),
                scoreNoise(a, b),
                scoreAge(a, b),
                scoreDiet(a, b),
                scorePresence(a, b),
                scoreHobbies(a, b)
        );

        // Somme ponderee
        double totalWeighted = breakdown.stream()
                .mapToDouble(d -> d.score() * d.weight())
                .sum();

        // Normalisation sur 100 (poids total = 0.95, on rescale)
        double total = (totalWeighted / ScoringWeights.TOTAL_WEIGHTS) * 100;

        // Cap a 100 pour eviter arrondi qui donnerait 100.01
        total = Math.min(100.0, total);

        // Arrondi a 1 decimale
        total = Math.round(total * 10) / 10.0;

        return new MatchingScore(total, breakdown);
    }

    // ============================================================
    // HARD FILTERS : elimination avant scoring
    // ============================================================

    /**
     * Verifie si 2 profils sont incompatibles au point d'etre elimines
     * avant meme de calculer le score.
     *
     * @return true si INCOMPATIBLE (a exclure), false si a scorer
     */
    public boolean isHardIncompatible(Profile a, Profile b) {
        // 1. Ville differente
        if (!a.getTargetCity().equalsIgnoreCase(b.getTargetCity())) {
            return true;
        }

        // 2. Pas d'overlap de budget
        if (a.getBudgetMaxEur() < b.getBudgetMinEur()) return true;
        if (b.getBudgetMaxEur() < a.getBudgetMinEur()) return true;

        // 3. Dates d'emmenagement trop eloignees
        long daysDiff = Math.abs(
                ChronoUnit.DAYS.between(a.getMoveInDate(), b.getMoveInDate())
        );
        if (daysDiff > ScoringWeights.MAX_MOVE_IN_DATE_DIFF_DAYS) return true;

        // 4. Genre prefere incompatible
        if (a.getPreferredGender() == PreferredGender.male && b.getGender() != Gender.male) {
            return true;
        }
        if (a.getPreferredGender() == PreferredGender.female && b.getGender() != Gender.female) {
            return true;
        }
        if (b.getPreferredGender() == PreferredGender.male && a.getGender() != Gender.male) {
            return true;
        }
        if (b.getPreferredGender() == PreferredGender.female && a.getGender() != Gender.female) {
            return true;
        }

        // 5. Tranche d'age preferee
        if (a.getPreferredAgeMin() != null && a.getPreferredAgeMax() != null) {
            if (b.getAge() < a.getPreferredAgeMin() || b.getAge() > a.getPreferredAgeMax()) {
                return true;
            }
        }
        if (b.getPreferredAgeMin() != null && b.getPreferredAgeMax() != null) {
            if (a.getAge() < b.getPreferredAgeMin() || a.getAge() > b.getPreferredAgeMax()) {
                return true;
            }
        }

        // 6. Type de logement incompatible
        if (a.getHousingType() != HousingType.any
                && b.getHousingType() != HousingType.any
                && a.getHousingType() != b.getHousingType()) {
            return true;
        }

        return false;
    }

    // ============================================================
    // DIMENSION 1 : BUDGET (20%)
    // ============================================================

    /**
     * Score du budget : plus l'overlap est grand par rapport aux plages, mieux c'est.
     * Ex: A=[500-800], B=[600-900] -> overlap=[600-800]=200 sur plages moyennes de 300
     *     ratio = 200/300 = 0.66 -> score 66
     */
    private DimensionResult scoreBudget(Profile a, Profile b) {
        int overlapMin = Math.max(a.getBudgetMinEur(), b.getBudgetMinEur());
        int overlapMax = Math.min(a.getBudgetMaxEur(), b.getBudgetMaxEur());

        double score;
        if (overlapMax < overlapMin) {
            // Pas d'overlap (deja gere par hard filter mais safety)
            score = 0.0;
        } else {
            int overlapSize = overlapMax - overlapMin;
            double avgRange = (
                    (a.getBudgetMaxEur() - a.getBudgetMinEur())
                            + (b.getBudgetMaxEur() - b.getBudgetMinEur())
            ) / 2.0;

            if (avgRange == 0) {
                // Les 2 users ont un budget fixe (min=max)
                score = 100.0;
            } else {
                double ratio = overlapSize / avgRange;
                score = Math.min(100.0, ratio * 100);
            }
        }

        return new DimensionResult("budget", score, ScoringWeights.WEIGHT_BUDGET, "Budget");
    }

    // ============================================================
    // DIMENSION 2 : SCHEDULE - horaires de sommeil (15%)
    // ============================================================

    private DimensionResult scoreSchedule(Profile a, Profile b) {
        int score = ScoringWeights.sleepCompatibility(a.getSleepSchedule(), b.getSleepSchedule());
        return new DimensionResult("schedule", score, ScoringWeights.WEIGHT_SCHEDULE, "Horaires");
    }

    // ============================================================
    // DIMENSION 3 : CLEANLINESS - proprete (12%)
    // ============================================================

    private DimensionResult scoreCleanliness(Profile a, Profile b) {
        int numA = ScoringWeights.LIFESTYLE_NUMERIC.get(a.getCleanliness());
        int numB = ScoringWeights.LIFESTYLE_NUMERIC.get(b.getCleanliness());
        int diff = Math.abs(numA - numB);

        int score = ScoringWeights.cleanlinessScore(diff);
        return new DimensionResult("cleanliness", score, ScoringWeights.WEIGHT_CLEANLINESS, "Proprete");
    }

    // ============================================================
    // DIMENSION 4 : SOCIAL - sociabilite (12%)
    // ============================================================

    private DimensionResult scoreSocial(Profile a, Profile b) {
        int score = ScoringWeights.socialCompatibility(a.getSocialLevel(), b.getSocialLevel());
        return new DimensionResult("social", score, ScoringWeights.WEIGHT_SOCIAL, "Sociabilite");
    }

    // ============================================================
    // DIMENSION 5 : SMOKING_PETS - fumeur & animaux (10%)
    // ============================================================

    private DimensionResult scoreSmokingPets(Profile a, Profile b) {
        // Score smoking
        double smokeScore = 100.0;

        // Si A ne fume pas, penaliser si B fume dedans
        if (a.getSmoking() == SmokingPreference.no_smoking) {
            if (b.getSmoking() == SmokingPreference.indoor_ok) smokeScore = 0;
            else if (b.getSmoking() == SmokingPreference.ok_outside) smokeScore = 60;
        }
        // Reciproque
        if (b.getSmoking() == SmokingPreference.no_smoking) {
            if (a.getSmoking() == SmokingPreference.indoor_ok) smokeScore = Math.min(smokeScore, 0);
            else if (a.getSmoking() == SmokingPreference.ok_outside) smokeScore = Math.min(smokeScore, 60);
        }

        // Score pets
        double petScore = 100.0;
        if (a.getPets() == PetsPreference.no_pets && b.getPets() == PetsPreference.has_pet) petScore = 0;
        if (b.getPets() == PetsPreference.no_pets && a.getPets() == PetsPreference.has_pet) petScore = 0;

        double avg = (smokeScore + petScore) / 2;
        return new DimensionResult("smoking_pets", avg, ScoringWeights.WEIGHT_SMOKING_PETS, "Fumeur & Animaux");
    }

    // ============================================================
    // DIMENSION 6 : NOISE - tolerance bruit vs frequence invites (8%)
    // ============================================================

    private DimensionResult scoreNoise(Profile a, Profile b) {
        int aNoise = ScoringWeights.LIFESTYLE_NUMERIC.get(a.getNoiseTolerance());
        int bNoise = ScoringWeights.LIFESTYLE_NUMERIC.get(b.getNoiseTolerance());
        int aGuests = ScoringWeights.LIFESTYLE_NUMERIC.get(a.getGuestsFrequency());
        int bGuests = ScoringWeights.LIFESTYLE_NUMERIC.get(b.getGuestsFrequency());

        // Mismatch : quand B invite beaucoup mais A ne tolere pas le bruit
        int mismatchAB = Math.max(0, bGuests - aNoise);
        int mismatchBA = Math.max(0, aGuests - bNoise);
        int maxMismatch = Math.max(mismatchAB, mismatchBA);

        int score = Math.max(0, 100 - (maxMismatch * 25));
        return new DimensionResult("noise", score, ScoringWeights.WEIGHT_NOISE, "Calme & Invites");
    }

    // ============================================================
    // DIMENSION 7 : AGE - difference d'age (3%)
    // ============================================================

    private DimensionResult scoreAge(Profile a, Profile b) {
        int diff = Math.abs(a.getAge() - b.getAge());
        int score = ScoringWeights.ageScore(diff);
        return new DimensionResult("age", score, ScoringWeights.WEIGHT_AGE, "Age");
    }

    // ============================================================
    // DIMENSION 8 : DIET - regime alimentaire (5%)
    // ============================================================

    private DimensionResult scoreDiet(Profile a, Profile b) {
        Diet dietA = a.getDiet();
        Diet dietB = b.getDiet();

        // Aucun renseignement -> score neutre
        if (dietA == null || dietB == null) {
            return new DimensionResult("diet", 70.0, ScoringWeights.WEIGHT_DIET, "Regime alimentaire");
        }

        // Meme regime -> parfait
        if (dietA == dietB) {
            return new DimensionResult("diet", 100.0, ScoringWeights.WEIGHT_DIET, "Regime alimentaire");
        }

        // Matrice de compatibilite (paires symetriques via EnumSet)
        Set<Diet> pair = EnumSet.of(dietA, dietB);
        double score = getDietPairScore(pair);

        return new DimensionResult("diet", score, ScoringWeights.WEIGHT_DIET, "Regime alimentaire");
    }

    private double getDietPairScore(Set<Diet> pair) {
        // Omnivore combinations
        if (pair.equals(EnumSet.of(Diet.omnivore, Diet.vegetarian))) return 70;
        if (pair.equals(EnumSet.of(Diet.omnivore, Diet.vegan)))      return 50;
        if (pair.equals(EnumSet.of(Diet.omnivore, Diet.halal)))      return 60;
        if (pair.equals(EnumSet.of(Diet.omnivore, Diet.kosher)))     return 60;
        if (pair.equals(EnumSet.of(Diet.omnivore, Diet.other)))      return 60;
        // Vegetarian combinations
        if (pair.equals(EnumSet.of(Diet.vegetarian, Diet.vegan)))    return 85;
        if (pair.equals(EnumSet.of(Diet.vegetarian, Diet.halal)))    return 75;
        if (pair.equals(EnumSet.of(Diet.vegetarian, Diet.kosher)))   return 75;
        if (pair.equals(EnumSet.of(Diet.vegetarian, Diet.other)))    return 70;
        // Vegan combinations
        if (pair.equals(EnumSet.of(Diet.vegan, Diet.halal)))         return 65;
        if (pair.equals(EnumSet.of(Diet.vegan, Diet.kosher)))        return 65;
        if (pair.equals(EnumSet.of(Diet.vegan, Diet.other)))         return 60;
        // Halal & Kosher
        if (pair.equals(EnumSet.of(Diet.halal, Diet.kosher)))        return 60;
        if (pair.equals(EnumSet.of(Diet.halal, Diet.other)))         return 70;
        if (pair.equals(EnumSet.of(Diet.kosher, Diet.other)))        return 70;

        return 60; // fallback
    }

    // ============================================================
    // DIMENSION 9 : PRESENCE - presence a la maison (5%)
    // ============================================================

    private DimensionResult scorePresence(Profile a, Profile b) {
        HomePresence presA = a.getHomePresence();
        HomePresence presB = b.getHomePresence();

        // Neutre si non renseigne
        if (presA == null || presB == null) {
            return new DimensionResult("presence", 70.0, ScoringWeights.WEIGHT_PRESENCE, "Presence");
        }

        double score;
        if (presA == presB) {
            score = 90.0; // Meme rythme
        } else if (
                (presA == HomePresence.mostly_home && presB == HomePresence.rarely_home)
                        || (presB == HomePresence.mostly_home && presA == HomePresence.rarely_home)
        ) {
            score = 80.0; // Un toujours la / un jamais la -> peut marcher
        } else {
            score = 75.0; // Complementaires
        }

        return new DimensionResult("presence", score, ScoringWeights.WEIGHT_PRESENCE, "Presence");
    }

    // ============================================================
    // DIMENSION 10 : HOBBIES - centres d'interet (5%)
    // ============================================================

    /**
     * Utilise la similarite de Jaccard : |A ∩ B| / |A ∪ B|
     * Ex: A=["sport","lecture"], B=["sport","cinema"]
     *     -> intersection={"sport"}, union={"sport","lecture","cinema"}
     *     -> jaccard = 1/3 = 0.33
     */
    private DimensionResult scoreHobbies(Profile a, Profile b) {
        List<String> hobbiesA = a.getHobbies();
        List<String> hobbiesB = b.getHobbies();

        // Un des deux n'a pas rempli -> score neutre
        if (hobbiesA == null || hobbiesA.isEmpty() || hobbiesB == null || hobbiesB.isEmpty()) {
            return new DimensionResult("hobbies", 60.0, ScoringWeights.WEIGHT_HOBBIES, "Centres d'interet");
        }

        Set<String> setA = new HashSet<>(hobbiesA);
        Set<String> setB = new HashSet<>(hobbiesB);

        // Intersection
        Set<String> intersection = new HashSet<>(setA);
        intersection.retainAll(setB);

        // Union
        Set<String> union = new HashSet<>(setA);
        union.addAll(setB);

        double jaccard = union.isEmpty() ? 0 : (double) intersection.size() / union.size();

        // Mapping jaccard -> score : 0 -> 40, 1.0 -> 100
        double score = 40 + (jaccard * 60);
        score = Math.round(score * 10) / 10.0; // 1 decimale

        return new DimensionResult("hobbies", score, ScoringWeights.WEIGHT_HOBBIES, "Centres d'interet");
    }
}